lib: utils: json: fix floating point encode and decode

EricNRS opened the pull request:
Add support for float and double floating types as primitives.

Fixes #59412
Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS committed:
lib: utils: json: fix floating point encode and decode

Add support for float and double floating types as primitives.

Fixes #59412
Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

ofirshe commented:
Hi @EricNRS,

Thanks again for your work on this! Your changes will make encoding and decoding floating-point primitives much easier. 

I wanted to point out that these two files currently use `JSON_TOK_FLOAT` in a way that will no longer work after your PR is merged since they rely on `json_obj_token`:
- [lwm2m_rw_senml_json.c](https://github.com/zephyrproject-rtos/zephyr/blob/e44a9523d4ed8dc01dbc9232431b8d6b095210eb/subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c)
- [lwm2m_rw_json.c](https://github.com/zephyrproject-rtos/zephyr/blob/e44a9523d4ed8dc01dbc9232431b8d6b095210eb/subsys/net/lib/lwm2m/lwm2m_rw_json.c)

Additionally, it looks like the prj.conf file in the JSON test is missing a newline at the end, which is causing a compliance check failure.

Best Regards.

EricNRS commented:
Thanks @ofirshe.

> I wanted to point out that these two files currently use `JSON_TOK_FLOAT` in a way that will no longer work after your PR is merged since they rely on `json_obj_token`:
> 
> * [lwm2m_rw_senml_json.c](https://github.com/zephyrproject-rtos/zephyr/blob/e44a9523d4ed8dc01dbc9232431b8d6b095210eb/subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c)
> * [lwm2m_rw_json.c](https://github.com/zephyrproject-rtos/zephyr/blob/e44a9523d4ed8dc01dbc9232431b8d6b095210eb/subsys/net/lib/lwm2m/lwm2m_rw_json.c)
> 

Thanks for that.  The code looks complicated and I didn't see any unit tests that referenced it.  One option is to leave the existing JSON_TOK_FLOAT behaviour as a raw type as I suspect that most people will want to use JSON_TOK_DOUBLE instead.

The other option is to get some assistance from recent contributors to help with the changes. 
@galak and @SeppoTakalo  - are able you able assist or point to someone that could help with the lwm2 JSON floating point refactor?

> Additionally, it looks like the prj.conf file in the JSON test is missing a newline at the end, which is causing a compliance check failure.

All fixed.

EricNRS commented:
The change itself is ready for review, but marking as draft until a fix for the following files have been identified:

- [lwm2m_rw_senml_json.c](https://github.com/zephyrproject-rtos/zephyr/blob/e44a9523d4ed8dc01dbc9232431b8d6b095210eb/subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c)
- [lwm2m_rw_json.c](https://github.com/zephyrproject-rtos/zephyr/blob/e44a9523d4ed8dc01dbc9232431b8d6b095210eb/subsys/net/lib/lwm2m/lwm2m_rw_json.c)

ofirshe commented on tests/lib/json/src/main.c:
```diff
@@ -425,6 +441,56 @@ ZTEST(lib_json_test, test_json_limits)
+
+	ret = json_obj_encode_buf(obj_double_descr, ARRAY_SIZE(obj_double_descr),
+				&td, buffer, sizeof(buffer));
+	zassert_equal(ret, 0, "Encoding float returned error");
```
```suggestion
	zassert_equal(ret, 0, "Encoding double returned error");
```

ofirshe commented on tests/lib/json/src/main.c:
```diff
@@ -425,6 +441,56 @@ ZTEST(lib_json_test, test_json_limits)
+
+	ret = json_obj_parse(encoded, sizeof(encoded) - 1, obj_double_descr,
+			     ARRAY_SIZE(obj_double_descr), &td_decoded);
+	zassert_equal(ret, 1, "Decoding float returned error");
```
```suggestion
	zassert_equal(ret, 1, "Decoding double returned error");
```

EricNRS replied:
Resolved in new changeset https://github.com/zephyrproject-rtos/zephyr/pull/86956

cjwinklhofer commented on include/zephyr/data/json.h:
```diff
@@ -44,6 +44,7 @@ enum json_tokens {
 	JSON_TOK_ENCODED_OBJ = '4',
 	JSON_TOK_INT64 = '5',
 	JSON_TOK_UINT64 = '6',
+	JSON_TOK_DOUBLE = '7',
```
To keep the previous float handling, you could use an additional token 'JSON_TOK_OPAQUE_FLOAT' that uses the previous implementation (e.g. _float_ascii_encode_ function).  Hence, the old parsing style would still be available and clients would only need to change the float token name, e.g. in _lwm2m_rw_json.c_ rename JSON_TOK_FLOAT to JSON_TOK_OPAQUE_FLOAT.

EricNRS replied:
I like that idea.

We still have users of this library that will need to change their code as well.  Guess I need to figure out how to add items to the migration guide - should I include the edits to https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-4.2.rst in this commit?

EricNRS replied:
Resolved in new changeset https://github.com/zephyrproject-rtos/zephyr/pull/86956

cjwinklhofer commented on lib/utils/json.c:
```diff
@@ -1088,11 +1153,37 @@ static int uint64_encode(const uint64_t *num, json_append_bytes_t append_bytes,
+static int double_encode(const double *num, json_append_bytes_t append_bytes,
 		      void *data)
 {
+	char buf[16 * sizeof(double)];
```
The buffer seems quite huge, I think it is sufficient to have 26 chars.
Sign + [0.] + digits + 'e+xxx' + '\0'

EricNRS replied:
Indeed, 128 bytes is a bit large.  I will fix that.

cjwinklhofer commented on lib/utils/json.c:
```diff
@@ -1088,11 +1153,37 @@ static int uint64_encode(const uint64_t *num, json_append_bytes_t append_bytes,
+
+static int float_encode(const float *num, json_append_bytes_t append_bytes, void *data)
+{
+	char buf[16 * sizeof(double)];
```
Buffer size could be smaller, see comment on decode_float.

EricNRS replied:
Resolved in new changeset https://github.com/zephyrproject-rtos/zephyr/pull/86956

cjwinklhofer commented on lib/utils/json.c:
Floats are encoded with '%g'. It would be nice to also decode the scientific format. You could extend the function `lexer_number` to allow scientific notation:

`if (isdigit(chr) != 0 || chr == '.' || chr == 'e' || chr == '+' || chr == '-') {`

sorry, did not find a way to comment on a specific line number.

EricNRS replied:
That is a really good catch.  I'll update the unit test to verify scientific notation as well.

cjwinklhofer commented:
Hi,
thanks for working on this JSON improvement. Independently, I also worked on the same feature and saw your pull-request shortly before submitting mine. Anyway, I will add some comments to your implementation.

Best
Christoph

cjwinklhofer commented on lib/utils/json.c:
```diff
@@ -1088,11 +1153,37 @@ static int uint64_encode(const uint64_t *num, json_append_bytes_t append_bytes,
 {
+	char buf[16 * sizeof(double)];
+	int ret;
 
```
Should nan and infinity result in an EINVAL error (same for float_encode):

```
if (isnan(*num) || isinf(*num)) {
        return -EINVAL;
}
```

EricNRS replied:
Resolved in new changeset https://github.com/zephyrproject-rtos/zephyr/pull/86956

cjwinklhofer commented:
For sake of completeness, I pushed my changes to github (without a pull-request). In case you want to use something, feel free and sorry, do not want to hijack the pull-request :)

https://github.com/cjwinklhofer/zephyr/tree/fix_json_floating_types

Thanks
Christoph

EricNRS commented:
> For sake of completeness, I pushed my changes to github (without a pull-request). In case you want to use something, feel free and sorry, do not want to hijack the pull-request :)
> 
> https://github.com/cjwinklhofer/zephyr/tree/fix_json_floating_types

You have addressed all of the comments here and I was not going to have time to get to this until later in the week, so I am happy :)  Go ahead and upload your changes as a PR which we can link to your PR from here and I will close this one.

For https://github.com/cjwinklhofer/zephyr/commit/cae9325d61d7761bf1e7c07a1cb020902a772fb4, I would suggest adding a note to the migration guide for the `JSON_TOK_OPAQUE_FLOAT` since there are probably a lot of people using it in out-of-tree builds and it makes the upgrade process easier.  I think this is the file to update:  https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-4.2.rst

EricNRS commented:
This PR has been replaced by the new PR https://github.com/zephyrproject-rtos/zephyr/pull/86956 from @cjwinklhofer.  Thanks Christoph!
